### PR TITLE
Add /api/v2/config endpoint

### PR DESCRIPTION
**Pull request by BogdanIrimie**
The topaz console is moving to the v2 config which contains a list of configs. To support this in topaz we are adding `/api/v2/config` endpoint which returns a list of configs and a flag that says if the configs are editable or not. In the case of topaz they are never editable that is why we set the `readOnly` flag to true

Sample output of the new endpoint
```json
{
    "readOnly": true,
    "authenticationEnabled": true,
    "configs": [
        {
            "configType": "auto",
            "name": "Topaz Config",
            "address": "https://localhost:4321/api/v2/config",
            "authorizerServiceUrl": "https://localhost:8383",
            "authorizerApiKey": "",
            "directoryServiceUrl": "https://localhost:9393",
            "directoryApiKey": "",
            "directoryTenantId": "",
            "directoryReaderServiceUrl": "https://localhost:9393",
            "directoryWriterServiceUrl": "https://localhost:9393",
            "directoryImporterServiceUrl": "https://localhost:9393",
            "directoryExporterServiceUrl": "https://localhost:9393",
            "directoryModelServiceUrl": "https://localhost:9393"
        }
    ]
}
```

Request example:
```
curl 'https://localhost:8080/api/v2/config' \
  -H 'content-type: application/json' \
  --compressed | jq
```

The API keys are returned only when the user is authenticated or if anonymous auth is enabled.

closes https://github.com/aserto-dev/workspace/issues/576

### CODE REVIEW

File: pkg/app/handlers/config.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+type ConsoleCfg struct {
+	AsertoDirectoryURL       string  `json:"asertoDirectoryUrl"`
+	AuthorizerServiceURL     string  `json:"authorizerServiceUrl"`
+	AuthorizerAPIKey         string  `json:"authorizerApiKey"`
+	DirectoryAPIKey          string  `json:"directoryApiKey"`
+	DirectoryTenantID        string  `json:"directoryTenantId"`
+	AsertoDirectoryReaderURL *string `json:"asertoDirectoryReaderUrl,omitempty"`
+	AsertoDirectoryWriterURL *string `json:"asertoDirectoryWriterUrl,omitempty"`
+	AsertoDirectoryModelURL  *string `json:"asertoDirectoryModelUrl,omitempty"`
```
Comment (gimmyxd) [inline pkg/app/handlers/config.go:17]: @BogdanIrimie could we create some symmetry between authoriser and directory?

Let me know if this looks good

```
AuthorizerServiceURL     string  `json:"authorizerServiceUrl"`
AuthorizerAPIKey         string  `json:"authorizerApiKey"`
DirectoryServiceURL       string  `json:"DirectoryServiceUrl"`
DirectoryAPIKey          string  `json:"directoryApiKey"`
DirectoryTenantID        string  `json:"directoryTenantId"`
DirectoryReaderServiceURL *string `json:"DirectoryReaderServiceUrl,omitempty"`
DirectoryWriterServiceURL *string `json:"DirectoryWriterServiceUrl,omitempty"`
DirectoryModelServiceURL  *string `json:"DirectoryModelServiceUrl,omitempty"`
```

**Comment by BogdanIrimie** [comment]
Added the `authenticatedUser` flag just so we know if the user was authenticated or not when the call was made. This makes it easy to distinguish between not receiving API keys because there are none or because the users was not authenticated.

File: pkg/app/auth/http.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		// if no API keys are defined or EnableAPIKey is not set, allow the request
+		options := authCfg.Options.ForPath(r.URL.Path)
+
+		if options.EnableAnonymous {
+			ctx := context.WithValue(r.Context(), handlers.AuthenticatedUser, true)
+			h.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+
+		if (len(authCfg.APIKeys) == 0) || !options.EnableAPIKey {
```
[the remark this answers is not in the document]
Comment (BogdanIrimie) [reply pkg/app/auth/http.go:23]: The is how it was previously set https://github.com/aserto-dev/topaz/blame/3b4c33713d301fe066bd079e1964329279a0e70d/pkg/app/auth/middleware.go#L95 and that is why I keep the same behaviour for `/api/v2/config`

File: pkg/app/auth/http.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		if authHeader == "" {
+			// auth header is not present =>  the user is unauthenticated and did not provide a token
+			ctx = context.WithValue(ctx, handlers.AuthenticatedUser, false)
+			h.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+
+		basicAPIKey, err := parseAuthHeader(authHeader, "basic")
+		if err != nil {
+			returnStatusUnauthorized(w, "Failed to parse basic auth header!", a.logger)
```
[the remark this answers is not in the document]
Comment (BogdanIrimie) [reply pkg/app/auth/http.go:41]: I keep it like this for consistency https://github.com/aserto-dev/topaz/blame/3b4c33713d301fe066bd079e1964329279a0e70d/pkg/app/auth/middleware.go#L101 but I can change it in both places.

File: pkg/app/auth/http.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ronenh) [reply pkg/app/auth/http.go:23]: We can fix it separately but this doesn't feel like desired behavior.

File: pkg/app/auth/http.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		}
+
+		if _, ok := authCfg.APIKeys[basicAPIKey]; ok {
+			ctx = context.WithValue(ctx, handlers.AuthenticatedUser, true)
+			h.ServeHTTP(w, r.WithContext(ctx))
+			return
+		}
+
+		// the user is not authenticated because the key they provided is incorrect
+		returnStatusUnauthorized(w, "The API key is invalid!", a.logger)
```
Comment (ronenh) [inline pkg/app/auth/http.go:52]: Here too. No need for exclamation points.

File: pkg/app/handlers/config.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	AuthenticatedUser = header.CtxKey("AuthenticatedUser")
+)
+
+type ConsoleCfg struct {
+	AuthorizerServiceURL        string  `json:"authorizerServiceUrl"`
+	AuthorizerAPIKey            string  `json:"authorizerApiKey"`
+	DirectoryServiceURL         string  `json:"directoryServiceUrl"`
+	DirectoryAPIKey             string  `json:"directoryApiKey"`
+	DirectoryTenantID           string  `json:"directoryTenantId"`
+	DirectoryReaderServiceURL   *string `json:"directoryReaderServiceUrl,omitempty"`
```
[the remark this answers is not in the document]
Comment (BogdanIrimie) [reply pkg/app/handlers/config.go]: They were defined like this previously https://github.com/aserto-dev/topaz/blob/3b4c33713d301fe066bd079e1964329279a0e70d/pkg/app/ui/handler.go#L15 and I don't see any reason not to make them strings. Changed them to be strings.

File: pkg/app/handlers/config.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+	"github.com/aserto-dev/header"
+)
+
+var (
+	AuthEnabled       = header.CtxKey("AuthEnabled")
+	AuthenticatedUser = header.CtxKey("AuthenticatedUser")
+)
+
+type ConsoleCfg struct {
```
[the remark this answers is not in the document]
Comment (BogdanIrimie) [reply pkg/app/handlers/config.go]: Renamed to TopazCfg

File: pkg/app/handlers/config.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+			DirectoryWriterServiceURL:   confServices.DirectoryWriterServiceURL,
+			DirectoryImporterServiceURL: confServices.DirectoryImporterServiceURL,
+			DirectoryExporterServiceURL: confServices.DirectoryExporterServiceURL,
+			DirectoryModelServiceURL:    confServices.DirectoryModelServiceURL,
+		}
+
+		cfgV2 := &TopazCfgV2{
+			Type:     "auto",
+			Name:     "Topaz Config",
+			Address:  "https://localhost:4321/api/v2/config",
```
Comment (gimmyxd) [inline pkg/app/handlers/config.go:96]: This seems to hardcode the Address. I think we need to provide the configured address for the console service here.

Comment (BogdanIrimie) [reply pkg/app/handlers/config.go]: Updated so we return the console URL e.g.

```json
{
    "readOnly": true,
    "configs": [
        {
            "configType": "auto",
            "name": "Topaz Config",
            "address": "https://localhost:8080/api/v2/config",
            "authorizerServiceUrl": "https://localhost:8383",
            "authorizerApiKey": "",
            "directoryServiceUrl": "https://localhost:9393",
            "directoryApiKey": "",
            "directoryTenantId": "",
            "directoryReaderServiceUrl": "https://localhost:9393",
            "directoryWriterServiceUrl": "https://localhost:9393",
            "directoryImporterServiceUrl": "https://localhost:9393",
            "directoryExporterServiceUrl": "https://localhost:9393",
            "directoryModelServiceUrl": "https://localhost:9393"
        }
    ]
}
```

**Review by ronenh — APPROVED** [review]

**Pull request merged**